packages/flutter/lib/src/widgets/widget_inspector.dart

```diff
@@ -31,6 +33,134 @@ typedef void _RegisterServiceExtensionCallback({
+///
+/// Requires that the render objects have already been painted successfully as
+/// part of the regular rendering pipeline.
+class _ScreenshotPaintingContext extends PaintingContext {
+  _ScreenshotPaintingContext._(
+    ContainerLayer containerLayer,
+    Rect estimatedBounds,
+  ) : super(containerLayer, estimatedBounds);
+
+  @override
+  void compositeLayerAtOffset(OffsetLayer layer, Offset offset) {
+    // Wrap the layer in a proxy layer to avoid detaching the layer from its
+    // existing parent or modifying its offset.
+    appendLayer(new _OffsetProxyLayer(layer, offset));
+  }
+
+  @override
+  void addLayer(Layer layer) {
+    // Wrap the layer in a proxy layer to avoid detaching the layer from its
+    // existing parent.
+    super.addLayer(new _ProxyLayer<Layer>(layer));
+  }
+
+  @override
+  PaintingContext createChildContext(ContainerLayer childLayer, Rect bounds) {
+    return new _ScreenshotPaintingContext._(childLayer, bounds);
+  }
+
+  /// Captures an image of the current state of [renderObject] and its children.
+  ///
+  /// The returned [ui.Image] has uncompressed raw RGBA bytes, will be offset
+  /// by the top-left corner of [renderBounds], and have dimensions equal to the
+  /// size of [renderBounds] multiplied by [pixelRatio].
+  ///
+  /// To use [toImage], the render object must have gone through the paint phase
+  /// (i.e. [debugNeedsPaint] must be false).
+  ///
+  /// The [pixelRatio] describes the scale between the logical pixels and the
+  /// size of the output image. It is independent of the
+  /// [window.devicePixelRatio] for the device, so specifying 1.0 (the default)
+  /// will give you a 1:1 mapping between logical pixels and the output pixels
+  /// in the image.
+  ///
+  /// [debugPaint] specifies whether the image should include the debug paint
+  /// for [renderObject]. Debug paint information is not included for the
+  /// children of [renderObject] so it is clear precisely which object the debug
+  /// paint information is for.
+  ///
+  /// See also:
+  ///
+  ///  * [RenderRepaintBoundary.toImage] for a similar API for [RenderObject]s
+  ///    that are repaint boundaries that can be used outside of the inspector.
+  ///  * [OffsetLayer.toImage] for a similar API at the layer level.
+  ///  * [dart:ui.Scene.toImage] for more information about the image returned.
+  static Future<ui.Image> toImage(
+    RenderObject renderObject,
+    Rect renderBounds, {
+    double pixelRatio = 1.0,
+    bool debugPaint = false,
+  }) {
+    final Float64List deviceTransform = new Float64List(16)
+      ..[0] = 1.0
+      ..[5] = 1.0
+      ..[10] = 1.0
+      ..[15] = 1.0;
+    final _ScreenshotPaintingContext context = new _ScreenshotPaintingContext._(
+      new TransformLayer(
+        transform: new Matrix4.fromFloat64List(deviceTransform),
+      ),
+      renderBounds,
+    );
+
+    context.paintChild(renderObject, Offset.zero);
+
+    if (debugPaint && debugPaint != debugPaintSizeEnabled) {
```

Comment:
this is probably more legible as `if (debugPaint && !debugPaintSizeEnabled)`

Reply:
done

Reply:
also added     
// The check that debugPaintSizeEnabled is false exists to ensure we only
// call debugPaint when it wasn't already called.
comment to clarify